_posts/2024/2024-02-06-about-caching-in-eo.md

```diff
@@ -0,0 +1,285 @@
+cache storage, and retrieval.
+To improve the flexibility for different data verification conditions, 
+the constructor of the `Cache` class  will accept a list of validations.
+Here's the corresponding code:
+
+```
+public class Cache {
+
+    private List<CacheValidation> validations;
+    
+    public Cache(final List<CacheValidation> cv) {
+        this.validations = cv;
+    }
+    
+    public Optional<XML> load(final Path source, final Path cache) {...};
+    
+    public void save(final Path cache, final Scalar<String> program, final Path relative) {...};
+}
+```
+
+The `List<CacheValidation>` represents a list of validations implemented from the `CacheValidation` interface.
+This interface defines the structure for validations within the `Cache` class.
+The `CacheValidation` interface has the only method ensuring that each validation contains a specific test condition.
+
+```
+public interface CacheValidation {
+    boolean validate(final Path source, final Path cache) throws IOException;
+}
+```
+
+2) In order to minimize disk access, we will utilize file paths represented by the `Path` class.
+By leveraging methods provided by the `Path` and `Files` classes,
+we can obtain essential information such as the file name and the time of the last modification.
+The file name plays a crucial role in locating the cached file within the directory,
+while the time of the last modification enables us to determine whether
+the source file is older or equal in age to the cached file.
+Given that the project build process in Maven is linear, 
+these conditions are deemed sufficient for our caching mechanism.
+
+
+3) Searching for a cached data will use the following conditions:
+  * `The source file` and `the cached file` should have same file name;
+  * Each goal involved caching should have both a cache directory and a directory of result files. 
+    The directory of result files corresponds to the directory of source files for the subsequent goal.
+  * The time of the last modification of the source file should be earlier or equal than cached file.
+
+
+Example: Let's consider an EO program named `program.eo`, which is executed for the first time.
+The cache of each goal will save the execution results in the cache directory and the result directory.
+When this program is run again without changes, these goal will receive data from the cache,
+without executing of task and rewriting of result.
+However, if we make changes to the `program.eo` file or `the source files` of the goals and execute again,
+the execution result of goals was overwritten in the directory of result files and the cache directory.
+This approach effectively protects the program from artificial changes during the build process.
+
+
+### Conclusion
+In this article, we explored various build systems and their caching methods.
+We were motivated to find an efficient caching approach for EO due to issues discovered during bug investigation.
```

Comment:
@Yanich96 To be honest, I don't understand why you "explored various build systems". This analysis is completely disconnected from the solution you propose. You observe high-level concepts in other build systems but fail to draw any conclusions from them. You merely mention their existence. Why? Why should we read about them? Where is the connection with the second part of the text?

Later, you just mentioned that we have some goals and some Mojos. Why? Why do we need to read about them? What is the purpose of this?

Furthermore, you only mention some classes. Why do they have this exact architecture? Why delve into such low-level details here? Perhaps it's better to suggest a high-level solution instead?

Please, pause for a moment and consider these questions:

What is the purpose of this blog post? Why are you writing it, and why should people read it?
At which level of abstraction do you want to discuss? High-level caching mechanisms like ccache, or low-level CacheValidation implementation?